Disable late attach to self by default

pdbain-ibm opened the pull request:
This addresses openj9 issue #126
fixes #126

Prevent a JVM from late attaching to itself unless the system property
"jdk.attach.allowAttachSelf" is set either to "true" or an empty string, per
the reference implementation. Enforce this in Java 9 only. Update existing
tests to enable this feature by default so they will continue to run. Add new
Java 9 tests to exercise this feature.

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

pdbain-ibm commented:
@DanHeidinga do you have any comments further to those in pull request #18?

DanHeidinga commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java:
```diff
@@ -0,0 +1,413 @@
+	/**
+	 * Reads any available lines of text from a stream without blocking, 
+	 * except to find the terminating newline.
+	 * This assumes the producer emits complete lines 
```
`ready()` returns true as soon as a `read()` operation won't block - i.e.: a single character is available.  This comment makes assumptions that aren't guaranteed by the Class libraries.

The usual way to do this is with separate threads that read from the buffered reader and append lines to some shared data structure.  

Is there something I'm missing here?

DanHeidinga replied:
Discussed deskside with Peter.  This isn't new code per se, its moving old code to a new location.  I'm willing to live it, especially as it's functioned for the past ~8 years.  

We should take a hard look at some of this code when the Java 8 vs 9 vs next test code is rationalized.

pdbain-ibm replied:
For new code, I agree with the suggestion. However, as discussed, this is cloned legacy code which we have been running for ~9 years.  Rewriting it now would add risk to this changeset.

I would be happy to re-do this code (and the Java8AndUp TargetManager) if you would open an issue.

pdbain-ibm commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java:
```diff
@@ -102,6 +102,15 @@
+	 * but parse it lazily because we rarely need the value.
+	 */
+	public final static String allowAttachSelf = 
+			com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties().getProperty("jdk.attach.allowAttachSelf");  //$NON-NLS-1$
```
Historical reasons.  Fixed.
Also cleaned up warnings on Java 5-era code.

pdbain-ibm commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java:
```diff
@@ -0,0 +1,413 @@
+	 * @param vmArgs VM arguments
+	 * @param appArgs Application arguments
+	 */
+	public TargetManager(String cmdName, String targetId, String displayName,
```
This code was cloned from the old Target manager, but fixed anyway.

pdbain-ibm commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/TestAttachAPI.java:
```diff
@@ -0,0 +1,104 @@
+ * This test must be invoked via testng.  Running the main method will return only the status of the attach API.
+ *
+ */
+@SuppressWarnings({"nls","boxing"})
```
Oops.  Missed that one.

pdbain-ibm commented:
Changeset updated.

pdbain-ibm commented:
Are any further changes required?

DanHeidinga commented:
I'm at JavaOne this week and likely won't get a chance to finish the review until next week.  I have no objections to one of the other committers completing the review / merge.

pdbain-ibm commented:
Are any further changes required?

pdbain-ibm commented:
@DanHeidinga Do you have any other comments on this pull request?

pshipton commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java:
```diff
@@ -50,7 +52,7 @@
 	 * Time delay before we give up trying to terminate the wait loop.
 	 */
-	static final long shutdownTimeoutMs = Integer.getInteger("com.ibm.tools.attach.shutdown_timeout", 10000); //$NON-NLS-1$
+	static final long shutdownTimeoutMs = Integer.getInteger("com.ibm.tools.attach.shutdown_timeout", 10000).longValue(); //$NON-NLS-1$
```
I'm not sure of the point of this change.

pdbain-ibm replied:
It avoids a warning about auto-unboxing.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,67 @@
+/*******************************************************************************
+ * Copyright (c) 2015, 2017 IBM Corp. and others
```
As this is new code the start date should be 2017

pdbain-ibm replied:
Fixed.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/TestAttachAPI.java:
```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2017 IBM Corp. and others
```
As this is new code the start date should be 2017.

pdbain-ibm replied:
Fixed.

pshipton commented on test/Java9andUp/src_current/org/openj9/test/java9AttachAPI/VmIdGetter.java:
```diff
@@ -0,0 +1,30 @@
+/* Copyright (c) 2015, 2017 IBM Corp. and others
```
As this is new code the start date should 2017.

pdbain-ibm replied:
Fixed.

pshipton commented on test/Java9andUp/src_latest/org/openj9/test/java9AttachAPI/VmIdGetter.java:
```diff
@@ -0,0 +1,30 @@
+/* Copyright (c) 2015, 2017 IBM Corp. and others
```
As this is new code the start date should 2017.

pdbain-ibm replied:
Fixed.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java:
```diff
@@ -0,0 +1,412 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2017 IBM Corp. and others
```
As this is new code the start date should be 2017.

pdbain-ibm replied:
Fixed.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,67 @@
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code is also Distributed under one or more Secondary Licenses,
```
These new files all have the incorrect license.

pdbain-ibm replied:
Fixed.  Let me know if I missed any.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java:
```diff
@@ -0,0 +1,412 @@
+		logger.debug("javaExec="+javaExec);
+		
+		/* These are the VM options to the parent process, which we wish to pass on to the child. */
+		String sideCar = System.getProperty("java.sidecar"); 
```
This ("java.sidecar") is obsolete, is this entire file an old copy of this helper?

pdbain-ibm replied:
Note that this is set by the playlist to indicate VM settings required by child processes (see test/Java8andUp/playlist.xml). Should I use a different system property?

pshipton replied:
Ok, I though it was a reference to something old and obsolete.

pshipton commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java:
```diff
@@ -102,10 +110,20 @@
 	private static String nameProperty;
 	private static String pidProperty;
 	private static int numberOfTargets;
+	/*[IF Sidecar19-SE]*/
```
Might be useful to add this code to Java 8 as well, but making the default true instead of false.

pdbain-ibm replied:
Good idea. Done

pshipton commented on jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties:
```diff
@@ -1262,3 +1262,4 @@ K0644="Caller-sensitive method called StackWalker.getCallerClass()"
 #java.lang.ClassLoader
 K0645="The given class loader name can't be empty."
 
+K0646="Late attach connection to self disabled. Set jdk.attach.allowAttachSelf=true"
```
Please update the preprocessor definition to match.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,68 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
```
Please update all the licenses to be the lastest version.

pshipton replied:
It appears the middle part of the licenses was updated, but not the ending part.

pshipton commented on test/Java9andUp/src/org/openj9/test/java9AttachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,68 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Please use 2017, 2017 in order to pass the copyright check.

pshipton commented:
@llxia any concerns with the test code?

pdbain-ibm commented:
@llxia or @sophia-guo Further to @pshipton 's comment above, is there a way I can share utility classes between the Java8AndUp and Java9AndUp projects?  That would simplify this PR considerably.  Thanks

pdbain-ibm commented:
Further to discussion with @smlambert et al, I will create a test utilities directory (under s separate pull request) to hold common code.  When that is complete, this pull request will be updated to use utilities in the common directory.

pdbain-ibm commented:
Pull request updated:
- removed the duplicate TargetManager and used the common copy in TestUtilities
- Renamed the test class and changed its package for compatibility with the Java 8 test code.
Otherwise it's the same.
@llxia could you please review the test code?

Thanks

smlambert commented on test/Java9andUp/src/org/openj9/test/attachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,70 @@
+			e.printStackTrace();
+			System.exit(ATTACH_ERROR_CODE);
+		}
+		System.exit(ATTACH_SELF_API_SUCCEEDED_CODE);
```
Do not think the System.exit calls are appropriate above, assert is likely what you should be doing.

pdbain-ibm replied:
I use the exit code to indicate the result of the late attach attempt to the parent test process, i.e.
```
			int status = target.waitFor();
			assertEquals(status, ATTACH_SELF_API_SUCCEEDED_CODE, "wrong exit status");
Again, will add a comment.

smlambert replied:
I should have stepped back and viewed the entire suite of attach API tests to see.  Comments will be useful thanks.

smlambert commented on test/Java9andUp/src/org/openj9/test/attachAPI/TestJava9AttachAPI.java:
```diff
@@ -0,0 +1,106 @@
+	}
+	
+	@Test
+	public void testAllowAttachSelfDisnable() {
```
Disable vs Disnable in method name

pdbain-ibm replied:
Oops.  Thanks for catching that.

smlambert commented on test/Java9andUp/src/org/openj9/test/attachAPI/TestJava9AttachAPI.java:
```diff
@@ -0,0 +1,106 @@
+				Arrays.asList(new String[] {ATTACH_ENABLE_PROPERTY,  ATTACH_SELF_ENABLE_PROPERTY+"=false"}), null);
+		try {
+			int status = target.waitFor();
+			assertEquals(status, ATTACH_SELF_IOEXCEPTION_CODE, "wrong exit status");
```
An opportunity for a better error message that wrong exit status (include exit status in the message would aid triage).

pdbain-ibm replied:
assertEquals() already does that, e.g.
wrong exit status expected [2] but found [1]

smlambert commented on test/Java9andUp/src_current/org/openj9/test/attachAPI/VmIdGetter.java:
```diff
@@ -0,0 +1,34 @@
+ *******************************************************************************/
+package org.openj9.test.attachAPI;
+/* This class is separated from the main test because we need two
+ * copies to accommodate old and new versions of ProcessHandle
```
@pshipton src_current is for supporting old class library version, which I thought we were trying to move from, instead of build up more test/support for.  Any guidance on this, why we keep adding more test material into this effort?

pdbain-ibm replied:
This needs to work on the vmfarm builds, which are stuck on an old (incompatible) class library.

pshipton replied:
This is adding a new test to IBM internal Java 9 b148 builds, correct? I agree its not necessary to add new testing. Existing testing either needs to be fixed or removed to keep the builds green. The b148 builds are obsolete and will be deleted once the equivalent testing is running on OpenJ9.

pdbain-ibm replied:
As discussed with the FV team, the tooling doesn't allow us to add the test only for the "latest" (i.e. post b148) JCL, so we will need to maintain 2 copies of some classes (as is done with the StackWalker tests) until we can kill b148.

llxia commented on test/Java9andUp/src/org/openj9/test/attachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,70 @@
+	public static void main(String[] args) {
+		try {
+			String myId = VmIdGetter.getVmId();
+			System.err.println("myId="+myId);
```
Should we should logger here?

pdbain-ibm replied:
Good question, but no, see above.

llxia commented on test/Java9andUp/src_latest/org/openj9/test/attachAPI/VmIdGetter.java:
```diff
@@ -0,0 +1,34 @@
+		return myId;
+	}
+
+}
```
We should put this file in test/Java9andUp/src to avoid duplication in both ssrc_latset and src_current.

pdbain-ibm replied:
Sadly no.  the ProcessHandle API changed from getPid() to pid();

pdbain-ibm commented on test/Java9andUp/src/org/openj9/test/attachAPI/SelfAttacher.java:
```diff
@@ -0,0 +1,70 @@
+	public static void main(String[] args) {
+		try {
+			String myId = VmIdGetter.getVmId();
+			System.err.println("myId="+myId);
```
The SelfAttacher is actually a child process.  It uses stderr to communicate its process ID (and hence VM ID) to the target manager.  I will add a comment to that effect.

pdbain-ibm commented:
Testing changes now.  Stay tuned...

pdbain-ibm commented:
Personal build passed.  @smlambert any more concerns?
Thanks

smlambert commented:
Jenkins test extended

pdbain commented:
failed tests due to known unrelated issue: org.openj9.test.vmArguments.VmArgumentTests.testCrNocr
